# Introduction to Jupyter

This is a Jupyter notebook. Notebooks are an environment to explore code, but also document your thought process.

Fundamentally, a notebook is a collection of cells.

Cells can contain code or text. Not just plain text, something called Markdown, which lets you add basic formatting. For example, to add a heading to a Markdown cell, use the # character, like this:

`# This is a heading` produces:

# This is a heading

This is not.


---

Now let's see some code. Jupyter notebooks aren't specific to Python and can be used for other programming languages, but for now we'll assume that "code" means Python.

We can run the code below by clicking the Run button or pressing `Ctrl+Enter` (or `Shift+Enter` which then moves to the cell below).

In [344]:
print("Hello world!")

Hello world!


---

# Introduction to `pandas`

`pandas` is one of the most important data science libraries in Python. It is used for:

- reading in and joining data from multiple sources
- exploring a dataset
- manipulating and reshaping data
- summary descriptive statistics
- visualisation (with the help of other libraries)

One of the most important things `pandas` introduces is new data types to work with. These are the `DataFrame` and the `Series`.

Before we can look at them, let's read in some data.

In [345]:
import pandas as pd

In [346]:
loans = pd.read_csv("./data/loans.csv")

When you read in data, it's a `DataFrame`. Think of it as a 2-D table.

In [347]:
type(loans)

pandas.core.frame.DataFrame

We can check the dimensions of our data:

In [348]:
loans.shape # (rows, columns) tuple

(42837, 13)

Or just get the length (number of rows):

In [349]:
len(loans)

42837

We can also inspect the column names:

In [350]:
loans.columns

Index(['id', 'loan_amnt', 'term', 'issue_d', 'int_rate', 'installment',
       'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership',
       'annual_inc', 'purpose'],
      dtype='object')

We can look at our data here in the Jupyter notebook:

In [351]:
loans.head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
0  130956066     3000.0   36 months  Mar-2018      7.34        93.10     A   
1  130968727     5000.0   36 months  Mar-2018     11.98       166.03     B   
2  130962380    20300.0   60 months  Mar-2018     17.47       509.66     D   
3  130966492    30000.0   36 months  Mar-2018     21.85      1143.39     D   
4  130945448     5000.0   36 months  Mar-2018      5.31       150.56     A   

  sub_grade                 emp_title emp_length home_ownership  annual_inc  \
0        A4          Scale Technician    9 years           RENT     52000.0   
1        B5            SDO Supervisor  10+ years            OWN     55000.0   
2        D1  Administrative Assistant   < 1 year           RENT     55000.0   
3        D5                   teacher  10+ years            OWN     57000.0   
4        A1                  Clerical    2 years           RENT     36610.0   

              purpose  
0      major_purchase  
1               other  
2         credit_card  
3  debt_consolidation  
4         credit_card

`pandas` automatically assigns row numbers to our data, called the index:

In [352]:
loans.index

RangeIndex(start=0, stop=42837, step=1)

The index of a `DataFrame` can be anything as long as the values are unique.

Looks like our columns are a mix of numeric and non-numeric types. We can verify this by looking at the data types of the columns:

In [353]:
loans.dtypes

id                  int64
loan_amnt         float64
term               object
issue_d            object
int_rate          float64
installment       float64
grade              object
sub_grade          object
emp_title          object
emp_length         object
home_ownership     object
annual_inc        float64
purpose            object
dtype: object

Anything that isn't a number, date, or boolean is an `object`. In `pandas` version 2, there is a dedicated string type for text, and categorical data also has its own type (even if it's text, this is a more targeted data type).

Before we start changing our columns' data types, let's see what a single column looks like.

The syntax for accessing a column in a `DataFrame` is similar to accessing an item in a dictionary:

In [354]:
loans["loan_amnt"]

0         3000.0
1         5000.0
2        20300.0
3        30000.0
4         5000.0
          ...   
42832    30000.0
42833    25000.0
42834    35000.0
42835    12300.0
42836     9000.0
Name: loan_amnt, Length: 42837, dtype: float64

A single column (and a single row) in `pandas` has the `Series` data type:

In [355]:
type(loans["loan_amnt"])

pandas.core.series.Series

You can select multiple columns by providing a list of column names:

In [356]:
loans[["id", "loan_amnt"]]

id  loan_amnt
0      130956066     3000.0
1      130968727     5000.0
2      130962380    20300.0
3      130966492    30000.0
4      130945448     5000.0
...          ...        ...
42832  130852914    30000.0
42833  130537040    25000.0
42834  130613820    35000.0
42835  130426636    12300.0
42836  130087499     9000.0

[42837 rows x 2 columns]

In [357]:
type(loans[["id", "loan_amnt"]])

pandas.core.frame.DataFrame

Remember: single columns/rows are `Series`, anything that's 2-D is a `DataFrame`.

Technically you can select a single column as a `DataFrame` by providing it in a list. This is useful for machine learning models that *need* a `DataFrame` as an input, even if it's a single column.

In [358]:
loans[["loan_amnt"]]

loan_amnt
0         3000.0
1         5000.0
2        20300.0
3        30000.0
4         5000.0
...          ...
42832    30000.0
42833    25000.0
42834    35000.0
42835    12300.0
42836     9000.0

[42837 rows x 1 columns]

In [359]:
type(loans[["loan_amnt"]])

pandas.core.frame.DataFrame

One way to change a column's data type is to use `astype`.

In [360]:
loans["loan_amnt"].astype(int)

0         3000
1         5000
2        20300
3        30000
4         5000
         ...  
42832    30000
42833    25000
42834    35000
42835    12300
42836     9000
Name: loan_amnt, Length: 42837, dtype: int64

If we look at the column again...

In [361]:
loans["loan_amnt"].dtype

dtype('float64')

It hasn't changed!

**Important**: in `pandas` most methods return *copies* of the data and do not change it.

You need to be *explicit* when you want your source data to change:

In [362]:
loans["loan_amnt"] = loans["loan_amnt"].astype(int)

loans["loan_amnt"].dtype

dtype('int64')

## Column transformations

A brief aside - why use `pandas` for data analysis?

One big reason is speed. `pandas` encourages **vectorised** operations, which means performing an operation on *all values in a column of data at once*.

Let's compare two approaches:

In [363]:
%%timeit

a = list(range(100_000))

new_list = []

for num in a:
    new_list.append(num**2)

20.2 ms ± 125 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [364]:
%%timeit

numbers = pd.Series(range(100_000))

# apply the operation to the entire Series at once
squares = numbers ** 2

137 μs ± 1.64 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


The biggest change to working with Python and working with `pandas` is getting rid of loops and working in a vectorised way.

#### Column operations with `pandas`

Let's calculate the loan installments as a % of the total loan amount:

In [365]:
loans["installment"] / loans["loan_amnt"]

0        0.031033
1        0.033206
2        0.025106
3        0.038113
4        0.030112
           ...   
42832    0.034685
42833    0.022234
42834    0.032230
42835    0.032230
42836    0.033206
Length: 42837, dtype: float64

Column operations return a `Series`, which we can assign to new column names to create new columns:

In [366]:
loans["installment_pct"] = loans["installment"] / loans["loan_amnt"]

loans.head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
0  130956066       3000   36 months  Mar-2018      7.34        93.10     A   
1  130968727       5000   36 months  Mar-2018     11.98       166.03     B   
2  130962380      20300   60 months  Mar-2018     17.47       509.66     D   
3  130966492      30000   36 months  Mar-2018     21.85      1143.39     D   
4  130945448       5000   36 months  Mar-2018      5.31       150.56     A   

  sub_grade                 emp_title emp_length home_ownership  annual_inc  \
0        A4          Scale Technician    9 years           RENT     52000.0   
1        B5            SDO Supervisor  10+ years            OWN     55000.0   
2        D1  Administrative Assistant   < 1 year           RENT     55000.0   
3        D5                   teacher  10+ years            OWN     57000.0   
4        A1                  Clerical    2 years           RENT     36610.0   

              purpose  installment_pct  
0      major_purchase         0.031033  
1               other         0.033206  
2         credit_card         0.025106  
3  debt_consolidation         0.038113  
4         credit_card         0.030112

You can also delete a column you don't like:

In [367]:
loans.drop(columns=["installment_pct"])

id  loan_amnt        term   issue_d  int_rate  installment  \
0      130956066       3000   36 months  Mar-2018      7.34        93.10   
1      130968727       5000   36 months  Mar-2018     11.98       166.03   
2      130962380      20300   60 months  Mar-2018     17.47       509.66   
3      130966492      30000   36 months  Mar-2018     21.85      1143.39   
4      130945448       5000   36 months  Mar-2018      5.31       150.56   
...          ...        ...         ...       ...       ...          ...   
42832  130852914      30000   36 months  Apr-2018     15.04      1040.55   
42833  130537040      25000   60 months  Apr-2018     11.98       555.86   
42834  130613820      35000   36 months  Apr-2018      9.92      1128.04   
42835  130426636      12300   36 months  Apr-2018      9.92       396.43   
42836  130087499       9000   36 months  Apr-2018     11.98       298.85   

      grade sub_grade                 emp_title emp_length home_ownership  \
0         A        A4          Scale Technician    9 years           RENT   
1         B        B5            SDO Supervisor  10+ years            OWN   
2         D        D1  Administrative Assistant   < 1 year           RENT   
3         D        D5                   teacher  10+ years            OWN   
4         A        A1                  Clerical    2 years           RENT   
...     ...       ...                       ...        ...            ...   
42832     C        C4         FINANCIAL MANAGER  10+ years       MORTGAGE   
42833     B        B5          General Manager      1 year       MORTGAGE   
42834     B        B2                Police Lt.  10+ years       MORTGAGE   
42835     B        B2  After-school coordinator    3 years       MORTGAGE   
42836     B        B5                       CNA  10+ years           RENT   

       annual_inc             purpose  
0         52000.0      major_purchase  
1         55000.0               other  
2         55000.0         credit_card  
3         57000.0  debt_consolidation  
4         36610.0         credit_card  
...           ...                 ...  
42832     64577.0         credit_card  
42833    110000.0  debt_consolidation  
42834    100000.0               other  
42835    112000.0  debt_consolidation  
42836     85000.0         credit_card  

[42837 rows x 13 columns]

In [368]:
loans.head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
0  130956066       3000   36 months  Mar-2018      7.34        93.10     A   
1  130968727       5000   36 months  Mar-2018     11.98       166.03     B   
2  130962380      20300   60 months  Mar-2018     17.47       509.66     D   
3  130966492      30000   36 months  Mar-2018     21.85      1143.39     D   
4  130945448       5000   36 months  Mar-2018      5.31       150.56     A   

  sub_grade                 emp_title emp_length home_ownership  annual_inc  \
0        A4          Scale Technician    9 years           RENT     52000.0   
1        B5            SDO Supervisor  10+ years            OWN     55000.0   
2        D1  Administrative Assistant   < 1 year           RENT     55000.0   
3        D5                   teacher  10+ years            OWN     57000.0   
4        A1                  Clerical    2 years           RENT     36610.0   

              purpose  installment_pct  
0      major_purchase         0.031033  
1               other         0.033206  
2         credit_card         0.025106  
3  debt_consolidation         0.038113  
4         credit_card         0.030112

Oh oh, it's still there!

Again, that's because `.drop()` doesn't modify the source data.

We can just overwrite our `DataFrame`:

In [369]:
loans = loans.drop(columns=["installment_pct"])
loans.head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
0  130956066       3000   36 months  Mar-2018      7.34        93.10     A   
1  130968727       5000   36 months  Mar-2018     11.98       166.03     B   
2  130962380      20300   60 months  Mar-2018     17.47       509.66     D   
3  130966492      30000   36 months  Mar-2018     21.85      1143.39     D   
4  130945448       5000   36 months  Mar-2018      5.31       150.56     A   

  sub_grade                 emp_title emp_length home_ownership  annual_inc  \
0        A4          Scale Technician    9 years           RENT     52000.0   
1        B5            SDO Supervisor  10+ years            OWN     55000.0   
2        D1  Administrative Assistant   < 1 year           RENT     55000.0   
3        D5                   teacher  10+ years            OWN     57000.0   
4        A1                  Clerical    2 years           RENT     36610.0   

              purpose  
0      major_purchase  
1               other  
2         credit_card  
3  debt_consolidation  
4         credit_card

When our data is the right type, you can perform type-specific operations:

In [370]:
loans["emp_title"] = loans["emp_title"].astype("string")

In [371]:
loans["emp_title"].str.lower()

0                scale technician
1                  sdo supervisor
2        administrative assistant
3                         teacher
4                        clerical
                   ...           
42832           financial manager
42833            general manager 
42834                  police lt.
42835    after-school coordinator
42836                         cna
Name: emp_title, Length: 42837, dtype: string

In [372]:
loans["emp_title"].str.lower().str.replace("manager", "mgr")

0                scale technician
1                  sdo supervisor
2        administrative assistant
3                         teacher
4                        clerical
                   ...           
42832               financial mgr
42833                general mgr 
42834                  police lt.
42835    after-school coordinator
42836                         cna
Name: emp_title, Length: 42837, dtype: string

We can also slice strings in a `pandas` column the same way as Python strings

(relevant `pandas` documentation: https://pandas.pydata.org/docs/user_guide/text.html)

In [373]:
loans["emp_title"].str[:5]

0        Scale
1        SDO S
2        Admin
3        teach
4        Cleri
         ...  
42832    FINAN
42833    Gener
42834    Polic
42835    After
42836      CNA
Name: emp_title, Length: 42837, dtype: string

For categorical data, such as the purpose of a loan, we can look at the unique values in a column:

In [374]:
loans["purpose"].unique()

array(['major_purchase', 'other', 'credit_card', 'debt_consolidation',
       'home_improvement', 'car', 'medical', 'house', 'small_business',
       'vacation', 'moving', 'renewable_energy', 'wedding'], dtype=object)

And we can count the number of unique values (either by counting the list above, or using `.nunique`:

In [375]:
loans["purpose"].nunique()

13

If you ever want to export the altered version of your data, you can do that!

In [376]:
# set index=False if you don't want to export the index as a separate column
loans.to_csv("loans_new.csv", index=False)

## Missing data

Typically we look at the number of missing records for each column:

In [496]:
loans.isnull().sum()

id                   0
loan_amnt            0
term                 0
issue_d              0
int_rate             0
installment          0
grade                0
sub_grade            0
emp_title            0
emp_length        3791
home_ownership       0
annual_inc           0
purpose              0
dtype: int64

You can also get that as a percentage:

In [378]:
loans.isnull().sum() / len(loans)

id                0.000000
loan_amnt         0.000000
term              0.000000
issue_d           0.000000
int_rate          0.000000
installment       0.000000
grade             0.000000
sub_grade         0.000000
emp_title         0.094965
emp_length        0.088498
home_ownership    0.000000
annual_inc        0.000000
purpose           0.000000
dtype: float64

We have a couple of options, we can:

- fill in the missing values with a placeholder like "unknown"
- drop rows with missing values

In [379]:
# .fillna() also doesn't change the underlying data!
loans["emp_title"] = loans["emp_title"].fillna("Unknown")

We can specify which columns to take into account when dropping data, and whether *all* of those columns need to be missing for us to drop a row, or *any* of them.

In [380]:
loans.dropna(subset=["emp_title", "emp_length"], how="all")

id  loan_amnt        term   issue_d  int_rate  installment  \
0      130956066       3000   36 months  Mar-2018      7.34        93.10   
1      130968727       5000   36 months  Mar-2018     11.98       166.03   
2      130962380      20300   60 months  Mar-2018     17.47       509.66   
3      130966492      30000   36 months  Mar-2018     21.85      1143.39   
4      130945448       5000   36 months  Mar-2018      5.31       150.56   
...          ...        ...         ...       ...       ...          ...   
42832  130852914      30000   36 months  Apr-2018     15.04      1040.55   
42833  130537040      25000   60 months  Apr-2018     11.98       555.86   
42834  130613820      35000   36 months  Apr-2018      9.92      1128.04   
42835  130426636      12300   36 months  Apr-2018      9.92       396.43   
42836  130087499       9000   36 months  Apr-2018     11.98       298.85   

      grade sub_grade                 emp_title emp_length home_ownership  \
0         A        A4          Scale Technician    9 years           RENT   
1         B        B5            SDO Supervisor  10+ years            OWN   
2         D        D1  Administrative Assistant   < 1 year           RENT   
3         D        D5                   teacher  10+ years            OWN   
4         A        A1                  Clerical    2 years           RENT   
...     ...       ...                       ...        ...            ...   
42832     C        C4         FINANCIAL MANAGER  10+ years       MORTGAGE   
42833     B        B5          General Manager      1 year       MORTGAGE   
42834     B        B2                Police Lt.  10+ years       MORTGAGE   
42835     B        B2  After-school coordinator    3 years       MORTGAGE   
42836     B        B5                       CNA  10+ years           RENT   

       annual_inc             purpose  
0         52000.0      major_purchase  
1         55000.0               other  
2         55000.0         credit_card  
3         57000.0  debt_consolidation  
4         36610.0         credit_card  
...           ...                 ...  
42832     64577.0         credit_card  
42833    110000.0  debt_consolidation  
42834    100000.0               other  
42835    112000.0  debt_consolidation  
42836     85000.0         credit_card  

[42837 rows x 13 columns]

<h1 style="color: #fcd805">Exercise: Exploratory Data Analysis with pandas</h1>

For the `pandas` exercises, you will gradually explore a new dataset of Kickstarter projects.

Kickstarter is a site that lets you crowdfund your project ideas. The dataset shows information about such projects including whether they succeeded or failed.

1. Read the file `kickstarter.csv.gz` from the `data` folder into a `pandas` `DataFrame` and inspect the data with the `.head` method.

Note: the `.gz` ending indicates this is a *zipped* CSV file. This greatly reduces the file size without losing any data, and the file can be read in exactly like a CSV file (no need to do anything about the fact that it's zipped, `pandas` will handle it).

In [381]:
ds = pd.read_csv("data/kickstarter.csv.gz")
ds


ID                                               name  \
0       1000002330                    The Songs of Adelaide & Abullah   
1       1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2       1000004038                                     Where is Hank?   
3       1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4       1000011046  Community Film Project: The Art of Neighborhoo...   
...            ...                                                ...   
378656   999976400  ChknTruk Nationwide Charity Drive 2014 (Canceled)   
378657   999977640                                          The Tribe   
378658   999986353  Walls of Remedy- New lesbian Romantic Comedy f...   
378659   999987933                           BioDefense Education Kit   
378660   999988282                  Nou Renmen Ayiti!  We Love Haiti!   

            subcategory      category currency    deadline     goal  \
0                Poetry    Publishing      GBP  2015-10-09   1000.0   
1        Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2        Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3                 Music         Music      USD  2012-04-16   5000.0   
4          Film & Video  Film & Video      USD  2015-08-29  19500.0   
...                 ...           ...      ...         ...      ...   
378656      Documentary  Film & Video      USD  2014-10-17  50000.0   
378657   Narrative Film  Film & Video      USD  2011-07-19   1500.0   
378658   Narrative Film  Film & Video      USD  2010-08-16  15000.0   
378659       Technology    Technology      USD  2016-02-13  15000.0   
378660  Performance Art           Art      USD  2011-08-16   2000.0   

                   launched  pledged     state  backers country  usd pledged  
0       2015-08-11 12:12:28      0.0    failed        0      GB          0.0  
1       2017-09-02 04:43:57   2421.0    failed       15      US        100.0  
2       2013-01-12 00:20:50    220.0    failed        3      US        220.0  
3       2012-03-17 03:24:11      1.0    failed        1      US          1.0  
4       2015-07-04 08:35:03   1283.0  canceled       14      US       1283.0  
...                     ...      ...       ...      ...     ...          ...  
378656  2014-09-17 02:35:30     25.0  canceled        1      US         25.0  
378657  2011-06-22 03:35:14    155.0    failed        5      US        155.0  
378658  2010-07-01 19:40:30     20.0    failed        1      US         20.0  
378659  2016-01-13 18:13:53    200.0    failed        6      US        200.0  
378660  2011-07-19 09:07:47    524.0    failed       17      US        524.0  

[378661 rows x 13 columns]

2. How many rows and columns are there?
3. A : 378661 rows , 13 columns

In [382]:
ds.shape

(378661, 13)

3. Check the data type of each column. Do any of them look incorrect?

In [383]:
ds.dtypes

ID               int64
name            object
subcategory     object
category        object
currency        object
deadline        object
goal           float64
launched        object
pledged        float64
state           object
backers          int64
country         object
usd pledged    float64
dtype: object

4. Are there any missing values? If so, what should be done about them?

In [384]:
ds.isnull().sum()

ID                0
name              4
subcategory       0
category          0
currency          0
deadline          0
goal              0
launched          0
pledged           0
state             0
backers           0
country           0
usd pledged    3797
dtype: int64

5. Create a new column to calculate the percentage of the goal that was achieved. This should be the amount pledged as a percentage of the goal.

In [385]:
ds["pct_goal"]= ds["pledged"] / ds["goal"]
ds


ID                                               name  \
0       1000002330                    The Songs of Adelaide & Abullah   
1       1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2       1000004038                                     Where is Hank?   
3       1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4       1000011046  Community Film Project: The Art of Neighborhoo...   
...            ...                                                ...   
378656   999976400  ChknTruk Nationwide Charity Drive 2014 (Canceled)   
378657   999977640                                          The Tribe   
378658   999986353  Walls of Remedy- New lesbian Romantic Comedy f...   
378659   999987933                           BioDefense Education Kit   
378660   999988282                  Nou Renmen Ayiti!  We Love Haiti!   

            subcategory      category currency    deadline     goal  \
0                Poetry    Publishing      GBP  2015-10-09   1000.0   
1        Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2        Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3                 Music         Music      USD  2012-04-16   5000.0   
4          Film & Video  Film & Video      USD  2015-08-29  19500.0   
...                 ...           ...      ...         ...      ...   
378656      Documentary  Film & Video      USD  2014-10-17  50000.0   
378657   Narrative Film  Film & Video      USD  2011-07-19   1500.0   
378658   Narrative Film  Film & Video      USD  2010-08-16  15000.0   
378659       Technology    Technology      USD  2016-02-13  15000.0   
378660  Performance Art           Art      USD  2011-08-16   2000.0   

                   launched  pledged     state  backers country  usd pledged  \
0       2015-08-11 12:12:28      0.0    failed        0      GB          0.0   
1       2017-09-02 04:43:57   2421.0    failed       15      US        100.0   
2       2013-01-12 00:20:50    220.0    failed        3      US        220.0   
3       2012-03-17 03:24:11      1.0    failed        1      US          1.0   
4       2015-07-04 08:35:03   1283.0  canceled       14      US       1283.0   
...                     ...      ...       ...      ...     ...          ...   
378656  2014-09-17 02:35:30     25.0  canceled        1      US         25.0   
378657  2011-06-22 03:35:14    155.0    failed        5      US        155.0   
378658  2010-07-01 19:40:30     20.0    failed        1      US         20.0   
378659  2016-01-13 18:13:53    200.0    failed        6      US        200.0   
378660  2011-07-19 09:07:47    524.0    failed       17      US        524.0   

        pct_goal  
0       0.000000  
1       0.080700  
2       0.004889  
3       0.000200  
4       0.065795  
...          ...  
378656  0.000500  
378657  0.103333  
378658  0.001333  
378659  0.013333  
378660  0.262000  

[378661 rows x 14 columns]

6. Drop the `usd pledged` column as it has some incorrect values in it.

In [386]:
ds = ds.drop(columns = "usd pledged")
ds


ID                                               name  \
0       1000002330                    The Songs of Adelaide & Abullah   
1       1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2       1000004038                                     Where is Hank?   
3       1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4       1000011046  Community Film Project: The Art of Neighborhoo...   
...            ...                                                ...   
378656   999976400  ChknTruk Nationwide Charity Drive 2014 (Canceled)   
378657   999977640                                          The Tribe   
378658   999986353  Walls of Remedy- New lesbian Romantic Comedy f...   
378659   999987933                           BioDefense Education Kit   
378660   999988282                  Nou Renmen Ayiti!  We Love Haiti!   

            subcategory      category currency    deadline     goal  \
0                Poetry    Publishing      GBP  2015-10-09   1000.0   
1        Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2        Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3                 Music         Music      USD  2012-04-16   5000.0   
4          Film & Video  Film & Video      USD  2015-08-29  19500.0   
...                 ...           ...      ...         ...      ...   
378656      Documentary  Film & Video      USD  2014-10-17  50000.0   
378657   Narrative Film  Film & Video      USD  2011-07-19   1500.0   
378658   Narrative Film  Film & Video      USD  2010-08-16  15000.0   
378659       Technology    Technology      USD  2016-02-13  15000.0   
378660  Performance Art           Art      USD  2011-08-16   2000.0   

                   launched  pledged     state  backers country  pct_goal  
0       2015-08-11 12:12:28      0.0    failed        0      GB  0.000000  
1       2017-09-02 04:43:57   2421.0    failed       15      US  0.080700  
2       2013-01-12 00:20:50    220.0    failed        3      US  0.004889  
3       2012-03-17 03:24:11      1.0    failed        1      US  0.000200  
4       2015-07-04 08:35:03   1283.0  canceled       14      US  0.065795  
...                     ...      ...       ...      ...     ...       ...  
378656  2014-09-17 02:35:30     25.0  canceled        1      US  0.000500  
378657  2011-06-22 03:35:14    155.0    failed        5      US  0.103333  
378658  2010-07-01 19:40:30     20.0    failed        1      US  0.001333  
378659  2016-01-13 18:13:53    200.0    failed        6      US  0.013333  
378660  2011-07-19 09:07:47    524.0    failed       17      US  0.262000  

[378661 rows x 13 columns]

7. Convert the `name` column to the `string` type.

In [387]:
ds["name"] = ds["name"].astype("string")
ds.dtypes

ID                      int64
name           string[python]
subcategory            object
category               object
currency               object
deadline               object
goal                  float64
launched               object
pledged               float64
state                  object
backers                 int64
country                object
pct_goal              float64
dtype: object

8. How many categories are there in the data, and what are they?
9. 15 - 'Publishing', 'Film & Video', 'Music', 'Food', 'Design', 'Crafts',
       'Games', 'Comics', 'Fashion', 'Theater', 'Art', 'Photography',
       'Technology', 'Dance', 'Journalism']

In [388]:
ds["category"].unique()
ds["state"].unique()

array(['failed', 'canceled', 'successful', 'live', 'undefined',
       'suspended'], dtype=object)

# Filtering

In `pandas` we can filter a dataset similarly to slicing a list, using square brackets `[]`. Within the square brackets should be a logical condition.

In `pandas`, this should be a `Series` of boolean values:

In [389]:
loan_filter = loans["loan_amnt"] > 30_000

loan_filter

0        False
1        False
2        False
3        False
4        False
         ...  
42832    False
42833    False
42834     True
42835    False
42836    False
Name: loan_amnt, Length: 42837, dtype: bool

Which we can then apply to the `DataFrame`:

In [390]:
loans[loan_filter].head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
19  130923172      30625   60 months  Mar-2018     16.01       744.91     C   
27  130492121      36250   36 months  Mar-2018     19.42      1336.50     D   
29  130925790      35000   36 months  Mar-2018     13.58      1189.09     C   
37  130872223      40000   60 months  Mar-2018      9.43       838.71     B   
47  130679755      38000   60 months  Mar-2018     10.41       815.08     B   

   sub_grade                  emp_title emp_length home_ownership  annual_inc  \
19        C5        Pharmacy Technician    2 years           RENT     25000.0   
27        D3                  Paralegal    7 years       MORTGAGE     45000.0   
29        C2  Aviation Safety Inspector    4 years       MORTGAGE    134315.0   
37        B1          Managing Director  10+ years           RENT    275000.0   
47        B3       Senior Web Developer    5 years            OWN     90000.0   

               purpose  
19  debt_consolidation  
27  debt_consolidation  
29  debt_consolidation  
37  debt_consolidation  
47  debt_consolidation

Or you can do it in one go:

In [391]:
loans[loans["loan_amnt"] > 30_000].head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
19  130923172      30625   60 months  Mar-2018     16.01       744.91     C   
27  130492121      36250   36 months  Mar-2018     19.42      1336.50     D   
29  130925790      35000   36 months  Mar-2018     13.58      1189.09     C   
37  130872223      40000   60 months  Mar-2018      9.43       838.71     B   
47  130679755      38000   60 months  Mar-2018     10.41       815.08     B   

   sub_grade                  emp_title emp_length home_ownership  annual_inc  \
19        C5        Pharmacy Technician    2 years           RENT     25000.0   
27        D3                  Paralegal    7 years       MORTGAGE     45000.0   
29        C2  Aviation Safety Inspector    4 years       MORTGAGE    134315.0   
37        B1          Managing Director  10+ years           RENT    275000.0   
47        B3       Senior Web Developer    5 years            OWN     90000.0   

               purpose  
19  debt_consolidation  
27  debt_consolidation  
29  debt_consolidation  
37  debt_consolidation  
47  debt_consolidation

Or you can use `.query()` for a more SQL-like filtering syntax:

In [392]:
loans.query("loan_amnt > 30000").head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
19  130923172      30625   60 months  Mar-2018     16.01       744.91     C   
27  130492121      36250   36 months  Mar-2018     19.42      1336.50     D   
29  130925790      35000   36 months  Mar-2018     13.58      1189.09     C   
37  130872223      40000   60 months  Mar-2018      9.43       838.71     B   
47  130679755      38000   60 months  Mar-2018     10.41       815.08     B   

   sub_grade                  emp_title emp_length home_ownership  annual_inc  \
19        C5        Pharmacy Technician    2 years           RENT     25000.0   
27        D3                  Paralegal    7 years       MORTGAGE     45000.0   
29        C2  Aviation Safety Inspector    4 years       MORTGAGE    134315.0   
37        B1          Managing Director  10+ years           RENT    275000.0   
47        B3       Senior Web Developer    5 years            OWN     90000.0   

               purpose  
19  debt_consolidation  
27  debt_consolidation  
29  debt_consolidation  
37  debt_consolidation  
47  debt_consolidation

More documentation on query and its options and limitations: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html

We can combine logical conditions too

- unlike in Python, we don't use `and` and `or`, we use `&` and `|`:
- each condition should be in its own brackets

In [393]:
loans[(loans["loan_amnt"] > 30_000) & (loans["int_rate"] > 15)].head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
19   130923172      30625   60 months  Mar-2018     16.01       744.91     C   
27   130492121      36250   36 months  Mar-2018     19.42      1336.50     D   
123  129579273      39725   60 months  Mar-2018     18.45      1018.51     D   
167  130928500      35000   60 months  Mar-2018     17.47       878.71     D   
179  130138254      40000   36 months  Mar-2018     15.04      1387.40     C   

    sub_grade                                 emp_title emp_length  \
19         C5                       Pharmacy Technician    2 years   
27         D3                                 Paralegal    7 years   
123        D2                                 Assembly      1 year   
167        D1  National Training Officer / Vice Preside    6 years   
179        C4                                   Unknown        NaN   

    home_ownership  annual_inc             purpose  
19            RENT     25000.0  debt_consolidation  
27        MORTGAGE     45000.0  debt_consolidation  
123       MORTGAGE     43000.0               other  
167       MORTGAGE    110000.0  debt_consolidation  
179       MORTGAGE         0.0  debt_consolidation

In [394]:
loans[(loans["loan_amnt"] > 30_000) | (loans["int_rate"] > 15)].head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
2   130962380      20300   60 months  Mar-2018     17.47       509.66     D   
3   130966492      30000   36 months  Mar-2018     21.85      1143.39     D   
8   130971296      12000   36 months  Mar-2018     15.04       416.22     C   
9   130373788      19000   60 months  Mar-2018     18.45       487.14     D   
11  130892912       1000   36 months  Mar-2018     17.47        35.89     D   

   sub_grade                 emp_title emp_length home_ownership  annual_inc  \
2         D1  Administrative Assistant   < 1 year           RENT     55000.0   
3         D5                   teacher  10+ years            OWN     57000.0   
8         C4                     Sales    3 years           RENT     43000.0   
9         D2                   TEACHER  10+ years       MORTGAGE     50000.0   
11        D1  Lead Sanitation Operator    2 years           RENT     51000.0   

               purpose  
2          credit_card  
3   debt_consolidation  
8   debt_consolidation  
9   debt_consolidation  
11         credit_card

We can also do type-specific filtering:

In [395]:
loans["emp_title"].str.lower().str.contains("teacher")

0        False
1        False
2        False
3         True
4        False
         ...  
42832    False
42833    False
42834    False
42835    False
42836    False
Name: emp_title, Length: 42837, dtype: boolean

In [396]:
teachers = loans[loans["emp_title"].str.lower().str.contains("teacher")]
teachers.head(10)

id  loan_amnt        term   issue_d  int_rate  installment grade  \
3    130966492      30000   36 months  Mar-2018     21.85      1143.39     D   
9    130373788      19000   60 months  Mar-2018     18.45       487.14     D   
77   130941096      30000   60 months  Mar-2018     24.84       877.73     E   
91   130537661      40000   36 months  Mar-2018      9.92      1289.19     B   
110  130929650       6000   36 months  Mar-2018     11.98       199.23     B   
180  130929752      35000   60 months  Mar-2018     14.07       815.66     C   
218  130934360      12000   36 months  Mar-2018     15.04       416.22     C   
254  130915978      14000   60 months  Mar-2018     20.39       373.96     D   
276  130373609      12000   36 months  Mar-2018     22.90       463.90     E   
298  130829408       4000   36 months  Mar-2018     23.87       156.66     E   

    sub_grade        emp_title emp_length home_ownership  annual_inc  \
3          D5          teacher  10+ years            OWN     57000.0   
9          D2          TEACHER  10+ years       MORTGAGE     50000.0   
77         E3          Teacher    2 years       MORTGAGE     81000.0   
91         B2          Teacher  10+ years           RENT     94000.0   
110        B5          Teacher  10+ years       MORTGAGE     80000.0   
180        C3          Teacher  10+ years       MORTGAGE     83000.0   
218        C4          Teacher    6 years            OWN     65000.0   
254        D4          teacher    9 years           RENT     75050.0   
276        E1  English Teacher   < 1 year            OWN     20000.0   
298        E2          Teacher  10+ years       MORTGAGE     60400.0   

                purpose  
3    debt_consolidation  
9    debt_consolidation  
77   debt_consolidation  
91     home_improvement  
110                 car  
180  debt_consolidation  
218  debt_consolidation  
254         credit_card  
276  debt_consolidation  
298  debt_consolidation

Any logical filters work, for example finding missing data:

In [397]:
loans[loans["emp_length"].isnull()].head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
36  130928745       2400   36 months  Mar-2018      9.43        76.81     B   
38  130865434      20000   60 months  Mar-2018     25.81       596.57     E   
39  130951563       5500   36 months  Mar-2018      7.34       170.69     A   
41  130946579       5000   36 months  Mar-2018      7.96       156.59     A   
55  130892809      25000   60 months  Mar-2018     17.47       627.65     D   

   sub_grade emp_title emp_length home_ownership  annual_inc  \
36        B1   Unknown        NaN       MORTGAGE    111000.0   
38        E4   Unknown        NaN           RENT     35000.0   
39        A4   Unknown        NaN            OWN     30696.0   
41        A5   Unknown        NaN           RENT     24600.0   
55        D1   Unknown        NaN       MORTGAGE     40000.0   

               purpose  
36    home_improvement  
38  debt_consolidation  
39  debt_consolidation  
41  debt_consolidation  
55    home_improvement

### `.loc` and `.iloc`

You can also do 2-D filtering (filtering rows and columns at the same time) using `.loc` or `.iloc`.

`.iloc` uses *integer indices*:

In [398]:
# first row
loans.iloc[0]

id                       130956066
loan_amnt                     3000
term                     36 months
issue_d                   Mar-2018
int_rate                      7.34
installment                   93.1
grade                            A
sub_grade                       A4
emp_title         Scale Technician
emp_length                 9 years
home_ownership                RENT
annual_inc                 52000.0
purpose             major_purchase
Name: 0, dtype: object

We can use slice notation to get multiple rows:

In [399]:
loans.iloc[0:5]

id  loan_amnt        term   issue_d  int_rate  installment grade  \
0  130956066       3000   36 months  Mar-2018      7.34        93.10     A   
1  130968727       5000   36 months  Mar-2018     11.98       166.03     B   
2  130962380      20300   60 months  Mar-2018     17.47       509.66     D   
3  130966492      30000   36 months  Mar-2018     21.85      1143.39     D   
4  130945448       5000   36 months  Mar-2018      5.31       150.56     A   

  sub_grade                 emp_title emp_length home_ownership  annual_inc  \
0        A4          Scale Technician    9 years           RENT     52000.0   
1        B5            SDO Supervisor  10+ years            OWN     55000.0   
2        D1  Administrative Assistant   < 1 year           RENT     55000.0   
3        D5                   teacher  10+ years            OWN     57000.0   
4        A1                  Clerical    2 years           RENT     36610.0   

              purpose  
0      major_purchase  
1               other  
2         credit_card  
3  debt_consolidation  
4         credit_card

And get 1 or more columns too:

In [400]:
loans.iloc[0:5, 0]

0    130956066
1    130968727
2    130962380
3    130966492
4    130945448
Name: id, dtype: int64

In contrast, `.loc` filters on the *actual* index (whether it's an integer or not).

So `.iloc[0]` always means the first row, whereas `.loc[0]` means the row with index (that's `pandas` index, not just list index) of 0.

In [401]:
loans.loc[0]

id                       130956066
loan_amnt                     3000
term                     36 months
issue_d                   Mar-2018
int_rate                      7.34
installment                   93.1
grade                            A
sub_grade                       A4
emp_title         Scale Technician
emp_length                 9 years
home_ownership                RENT
annual_inc                 52000.0
purpose             major_purchase
Name: 0, dtype: object

Column filtering with `.loc` is done by *name*:

In [402]:
loans.loc[0:5, "loan_amnt"]

0     3000
1     5000
2    20300
3    30000
4     5000
5    15000
Name: loan_amnt, dtype: int64

And we can slice column names!

In [403]:
loans.loc[0:5, "loan_amnt":"int_rate"]

loan_amnt        term   issue_d  int_rate
0       3000   36 months  Mar-2018      7.34
1       5000   36 months  Mar-2018     11.98
2      20300   60 months  Mar-2018     17.47
3      30000   36 months  Mar-2018     21.85
4       5000   36 months  Mar-2018      5.31
5      15000   36 months  Mar-2018      6.07

Finally, `.loc` allows boolean filtering:

In [404]:
loans.loc[loans["int_rate"] < 5.5, "int_rate"]

4        5.31
81       5.31
183      5.31
220      5.31
239      5.31
         ... 
42666    5.31
42708    5.31
42772    5.31
42802    5.31
42817    5.31
Name: int_rate, Length: 912, dtype: float64

## Descriptive statistics

`pandas` has built-in methods for getting some statistical summaries of our data:

In [405]:
loans.describe()

id     loan_amnt      int_rate   installment    annual_inc
count  4.283700e+04  42837.000000  42837.000000  42837.000000  4.283700e+04
mean   1.321465e+08  15367.714242     13.864898    458.324981  7.864097e+04
std    4.401853e+06  10231.349824      5.602657    295.928644  7.922587e+04
min    7.010232e+07   1000.000000      5.310000     30.120000  0.000000e+00
25%    1.286246e+08   7200.000000      9.920000    233.100000  4.600000e+04
50%    1.313419e+08  12250.000000     13.060000    375.880000  6.500000e+04
75%    1.349583e+08  21000.000000     17.470000    631.130000  9.500000e+04
max    1.452886e+08  40000.000000     30.990000   1628.080000  9.300000e+06

By default, only numeric values are shown but we can change that:

In [406]:
loans.describe(include="all")

id     loan_amnt        term   issue_d      int_rate  \
count   4.283700e+04  42837.000000       42837     42837  42837.000000   
unique           NaN           NaN           2        12           NaN   
top              NaN           NaN   36 months  Jan-2018           NaN   
freq             NaN           NaN       30319      7298           NaN   
mean    1.321465e+08  15367.714242         NaN       NaN     13.864898   
std     4.401853e+06  10231.349824         NaN       NaN      5.602657   
min     7.010232e+07   1000.000000         NaN       NaN      5.310000   
25%     1.286246e+08   7200.000000         NaN       NaN      9.920000   
50%     1.313419e+08  12250.000000         NaN       NaN     13.060000   
75%     1.349583e+08  21000.000000         NaN       NaN     17.470000   
max     1.452886e+08  40000.000000         NaN       NaN     30.990000   

         installment  grade sub_grade emp_title emp_length home_ownership  \
count   42837.000000  42837     42837     42837      39046          42837   
unique           NaN      7        35     17973         11              4   
top              NaN      C        B5   Unknown  10+ years       MORTGAGE   
freq             NaN  11745      2526      4068      14187          21155   
mean      458.324981    NaN       NaN       NaN        NaN            NaN   
std       295.928644    NaN       NaN       NaN        NaN            NaN   
min        30.120000    NaN       NaN       NaN        NaN            NaN   
25%       233.100000    NaN       NaN       NaN        NaN            NaN   
50%       375.880000    NaN       NaN       NaN        NaN            NaN   
75%       631.130000    NaN       NaN       NaN        NaN            NaN   
max      1628.080000    NaN       NaN       NaN        NaN            NaN   

          annual_inc             purpose  
count   4.283700e+04               42837  
unique           NaN                  13  
top              NaN  debt_consolidation  
freq             NaN               21949  
mean    7.864097e+04                 NaN  
std     7.922587e+04                 NaN  
min     0.000000e+00                 NaN  
25%     4.600000e+04                 NaN  
50%     6.500000e+04                 NaN  
75%     9.500000e+04                 NaN  
max     9.300000e+06                 NaN

Individual columns can also be summarised:

In [407]:
loans["loan_amnt"].mean()

np.float64(15367.714242360576)

In [408]:
loans["loan_amnt"].median()

np.float64(12250.0)

You can also use `.agg` to request multiple summaries:

In [409]:
loans["loan_amnt"].agg(["min", "max"])

min     1000
max    40000
Name: loan_amnt, dtype: int64

<h1 style="color: #fcd805">Exercise: Filtering and Descriptive Statistics</h1>

We're going to continue working on the Kickstarter data from the previous exercise.

1. How many projects are in the Music category?
2. 51918

In [410]:
len(ds.query("category=='Music'"))

51918

2. How many projects in the Music category *succeeded*?
3. 24197

In [411]:
len(ds.query("category=='Music' and state=='successful'"))

24197

3. How many projects in the Music category contain the word "song"?
4. 479

In [412]:
len(ds.query("category=='Music' and name.str.lower().str.contains('song')"))

1612

4. How many projects are in the Music and Film & Video categories in total?
5. 115503

In [413]:
len(ds.query("category=='Music' or category == 'Film & Video' "))

115503

5. What are the smallest and biggest goals in the dataset?
6. min    1.000000e-02
max    1.000000e+08
Name: goal, dtype: float64

In [414]:
ds["goal"].agg(["min","max"])


min    1.000000e-02
max    1.000000e+08
Name: goal, dtype: float64

6. What is the average number of backers a project received?
7. np.float64(105.61747578969052)

In [415]:
ds["backers"].mean()
ds[ds["state"]=="successful"]["backers"].mean()



np.float64(263.92136223834694)

In [416]:
ds[ds["state"]=="successful"]["backers"].mean()

np.float64(263.92136223834694)

7. What is the average number of backers that *successful* projects received?

_Hint: Think about the order of operations of how to answer this. What do you need to do first?_

# Sorting and aggregating

We can sort our data:

In [417]:
loans.sort_values("grade").head()

id  loan_amnt        term   issue_d  int_rate  installment  \
0      130956066       3000   36 months  Mar-2018      7.34        93.10   
27282  142170251      40000   36 months  Oct-2018      6.67      1229.06   
27280  142061144       1800   36 months  Oct-2018      8.46        56.79   
8868   128565824       3300   36 months  Feb-2018      6.08       100.52   
27279  142153417       5000   36 months  Oct-2018      8.46       157.75   

      grade sub_grade           emp_title emp_length home_ownership  \
0         A        A4    Scale Technician    9 years           RENT   
27282     A        A2          Jr partner  10+ years       MORTGAGE   
27280     A        A5  Product Specialist     1 year       MORTGAGE   
8868      A        A2             Unknown        NaN       MORTGAGE   
27279     A        A5             Unknown        NaN            OWN   

       annual_inc             purpose  
0         52000.0      major_purchase  
27282    300000.0  debt_consolidation  
27280     62000.0               other  
8868      85000.0             medical  
27279     35000.0         credit_card

We can sort by multiple values:

In [418]:
loans.sort_values(["grade", "sub_grade"]).head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
4    130945448       5000   36 months  Mar-2018      5.31       150.56     A   
81   130803047      15000   36 months  Mar-2018      5.31       451.66     A   
183  130825815      40000   36 months  Mar-2018      5.31      1204.42     A   
220  130300823       5000   36 months  Mar-2018      5.31       150.56     A   
239  130665575       4000   36 months  Mar-2018      5.31       120.45     A   

    sub_grade                       emp_title emp_length home_ownership  \
4          A1                        Clerical    2 years           RENT   
81         A1                         Unknown        NaN       MORTGAGE   
183        A1      Senior Mechanical Engineer    2 years       MORTGAGE   
220        A1  Senior Creative Director, Copy    5 years       MORTGAGE   
239        A1                             CFO    5 years       MORTGAGE   

     annual_inc             purpose  
4       36610.0         credit_card  
81      49000.0  debt_consolidation  
183    139000.0    home_improvement  
220    150000.0               house  
239    200000.0    home_improvement

We can change the direction of sorting either for a single column, or different sort orders for multiple columns:

In [419]:
loans.sort_values("grade", ascending=False).head()

id  loan_amnt        term   issue_d  int_rate  installment  \
22109  138139457      16000   60 months  Aug-2018     30.79       525.45   
23646  137213184      38700   60 months  Jul-2018     30.84      1272.12   
34695  133048019      28000   60 months  May-2018     30.79       919.53   
27623  141282807      17925   60 months  Oct-2018     30.79       588.67   
24570  136206447      14500   60 months  Jul-2018     30.99       477.98   

      grade sub_grade              emp_title emp_length home_ownership  \
22109     G        G1   Fixed income trading   < 1 year       MORTGAGE   
23646     G        G2  Business Front Office    2 years       MORTGAGE   
34695     G        G1              SALES Rep    2 years            OWN   
27623     G        G1             ACCOUNTANT  10+ years       MORTGAGE   
24570     G        G5                 Driver    2 years           RENT   

       annual_inc             purpose  
22109    180000.0    home_improvement  
23646     35800.0  debt_consolidation  
34695     70000.0  debt_consolidation  
27623     64000.0               other  
24570     55000.0      major_purchase

In [420]:
loans.sort_values(["grade", "sub_grade"], ascending=[True, False]).head()

id  loan_amnt        term   issue_d  int_rate  installment grade  \
12  130874297       3500   36 months  Mar-2018      7.96       109.62     A   
20  130961445      24000   36 months  Mar-2018      7.96       751.63     A   
32  130872775      10000   36 months  Mar-2018      7.96       313.18     A   
41  130946579       5000   36 months  Mar-2018      7.96       156.59     A   
58  130841128      20000   36 months  Mar-2018      7.96       626.36     A   

   sub_grade              emp_title emp_length home_ownership  annual_inc  \
12        A5      Vehicle Inventory    4 years       MORTGAGE     57000.0   
20        A5  Adminstrative Analyst  10+ years           RENT    109000.0   
32        A5                Manager    5 years       MORTGAGE     67000.0   
41        A5                Unknown        NaN           RENT     24600.0   
58        A5                  owner  10+ years            OWN     90000.0   

               purpose  
12    home_improvement  
20    home_improvement  
32  debt_consolidation  
41  debt_consolidation  
58         credit_card

For categorical data, we can look at the frequencies of the values using `value_counts`

In [421]:
loans["home_ownership"].value_counts()

home_ownership
MORTGAGE    21155
RENT        15899
OWN          5771
ANY            12
Name: count, dtype: int64

Or as a percentage:

In [422]:
loans["home_ownership"].value_counts(normalize=True)

home_ownership
MORTGAGE    0.493849
RENT        0.371151
OWN         0.134720
ANY         0.000280
Name: proportion, dtype: float64

The default for this table is to sort values in descending order, but we can sort by the labels (which is the index):

In [423]:
loans["home_ownership"].value_counts().sort_index()

home_ownership
ANY            12
MORTGAGE    21155
OWN          5771
RENT        15899
Name: count, dtype: int64

### Groupby

Like in SQL and other data-specific languages, we can use `groupby` to create subsets of our data and aggregate them separately:

In [424]:
loans.groupby("home_ownership")

That doesn't do anything yet because we haven't chosen how to summarise each group!

In [425]:
loans.groupby("home_ownership")["loan_amnt"].median()

home_ownership
ANY          8600.0
MORTGAGE    15000.0
OWN         12000.0
RENT        11000.0
Name: loan_amnt, dtype: float64

The default here is to order by label (index) but we can sort on the values:

In [426]:
loans.groupby("home_ownership")["loan_amnt"].median().sort_values(ascending=False)

home_ownership
MORTGAGE    15000.0
OWN         12000.0
RENT        11000.0
ANY          8600.0
Name: loan_amnt, dtype: float64

We can provide multiple columns to summarise:

In [427]:
loans.groupby("home_ownership")[["loan_amnt", "int_rate"]].median()

loan_amnt  int_rate
home_ownership                     
ANY                8600.0     16.73
MORTGAGE          15000.0     12.61
OWN               12000.0     12.73
RENT              11000.0     13.59

Or multiple aggregations on a single column:

In [428]:
loans.groupby("home_ownership")["loan_amnt"].agg(["min", "median", "max"])

min   median    max
home_ownership                      
ANY             2200   8600.0  25000
MORTGAGE        1000  15000.0  40000
OWN             1000  12000.0  40000
RENT            1000  11000.0  40000

Or multiple aggregations for multiple columns!

In [429]:
loans.groupby("home_ownership")[["loan_amnt", "int_rate"]].agg(["min", "median", "max"])

loan_amnt                 int_rate              
                     min   median    max      min median    max
home_ownership                                                 
ANY                 2200   8600.0  25000    10.47  16.73  29.69
MORTGAGE            1000  15000.0  40000     5.31  12.61  30.99
OWN                 1000  12000.0  40000     5.31  12.73  30.94
RENT                1000  11000.0  40000     5.31  13.59  30.99

<h1 style="color: #fcd805">Exercise: Sorting and Aggregating</h1>

Back to the Kickstarter data.

1. What is the **total** amount pledged for songs *by category*?

You want to end up with a dataset of one line per category, showing the total pledged amount for each category.

In [430]:
ds.head(2)
ds.groupby("category")["pledged"].sum()

category
Art             1.015470e+08
Comics          7.464365e+07
Crafts          1.776030e+07
Dance           1.390693e+07
Design          8.154909e+08
Fashion         1.494227e+08
Film & Video    4.045744e+08
Food            1.313787e+08
Games           7.703319e+08
Journalism      1.530200e+07
Music           2.072948e+08
Photography     3.950123e+07
Publishing      1.450902e+08
Technology      7.356088e+08
Theater         4.471301e+07
Name: pledged, dtype: float64

2. What is the breakdown of the state of projects? That is, how many have failed, succeeded etc.? Calculate the answer both as absolute numbers and percentages.

In [431]:
ds.head(2)
#ks=ds["state"].fillna("NA")
ds["state"].value_counts(normalize=True)
#ds.groupby("state")["state"].count()

state
failed        0.522153
successful    0.353762
canceled      0.102411
undefined     0.009407
live          0.007392
suspended     0.004875
Name: proportion, dtype: float64

3. Which category has the highest *average* pledged amount?

In [432]:
ds.head(2)
ds.groupby("category")["pledged"].mean().reset_index().sort_values("pledged",ascending = False).iloc[0]

category          Design
pledged     27119.751279
Name: 4, dtype: object

4. Find the most expensive (i.e. highest goal) project in the Photography category.

In [433]:
ds[ds["category"]=="Photography"].sort_values("goal", ascending = False ).head(1)
#ds[ds["category"]=="Photography"]["goal"].idxmax()

ID                                           name  subcategory  \
5076  1025947904  Long island city new york art book (Canceled)  Photography   

         category currency    deadline        goal             launched  \
5076  Photography      USD  2013-11-07  10000000.0  2013-10-08 23:16:26   

      pledged     state  backers country  pct_goal  
5076      0.0  canceled        0      US       0.0

5. Find the project in the Food category with the highest number of backers.

In [434]:
ds[ds["category"]=="Food"].sort_values("backers", ascending = False ).head(1)

ID               name subcategory category currency  \
246496  323562295  Misen: Cook Sharp        Food     Food      USD   

          deadline     goal             launched     pledged       state  \
246496  2015-10-22  25000.0  2015-09-22 15:02:12  1083344.11  successful   

        backers country   pct_goal  
246496    13116      US  43.333764

6. **BONUS** Find the project with the longest name.

_Hint: figure out how to calculate the length of the names first!_

In [435]:
ds.loc[ds["name"].str.len().sort_values(ascending = False ).idxmax(),"name"]

'BENEATH - Publish the book â\x80¢ Visit another world â\x80¢ Lose your mind â\x80¢ Cringe in fear (Canceled)'

# Combining data

`pandas` supports much more than just CSV files. It can connect to many different data sources.

Once data is read into `pandas` it is always a `DataFrame` regardless of where it came from. This is one of the strengths of `pandas` and it means you can combine data *from different sources*.

Let's see how we could connect to a SQL database.

In [436]:
import sqlite3

conn = sqlite3.connect("./data/movies.sqlite")

type(conn)

sqlite3.Connection

There are 3 tables in this database:

- IMDB (a list of films)
- earning (the amount of money grossed by each film)
- genre (the genre of each film)

We can use `pandas` to directly run a SQL query on our database and save the data as a `DataFrame`.

In [437]:
films = pd.read_sql("""
SELECT
    *
FROM
    IMDB
""", conn)

type(films)

pandas.core.frame.DataFrame

In [438]:
films.head()

Movie_id                    Title  Rating  TotalVotes MetaCritic  \
0    36809  12 Years a Slave (2013)     8.1      496092         96   
1    30114         127 Hours (2010)     7.6      297075         82   
2    37367             50/50 (2011)     7.7      283935         72   
3    49473        About Time (2013)     7.8      225412              
4    14867             Amour (2012)     7.9       76121         94   

       Budget  Runtime CVotes10 CVotes09 CVotes08  ... Votes3044 Votes3044M  \
0  20000000.0  134 min    75556   126223   161460  ...       8.0        7.9   
1  18000000.0   94 min    28939    44110    98845  ...       7.5        7.5   
2   8000000.0  100 min    28304    47501    99524  ...       7.6        7.6   
3  12000000.0  123 min    38556    43170    70850  ...       7.6        7.6   
4   8900000.0  127 min    11093    15944    22942  ...       7.7        7.7   

  Votes3044F Votes45A Votes45AM Votes45AF VotesIMDB Votes1000 VotesUS VotesnUS  
0        8.0      7.8       7.8       8.1       8.0       7.7     8.3      8.0  
1        7.5      7.3       7.3       7.5       7.6       7.0     7.7      7.6  
2        7.6      7.4       7.4       7.5       7.4       7.0     7.9      7.6  
3        7.7      7.6       7.5       7.8       7.7       6.9     7.8      7.7  
4        7.9      7.9       7.8       8.1       6.6       7.2     7.9      7.8  

[5 rows x 52 columns]

Let's also select the genre data

In [439]:
genres = pd.read_sql("""
SELECT
    *
FROM
    genre
""", conn)

genres.head()

Movie_id      genre
0    36809  Biography
1    30114  Adventure
2    37367     Comedy
3    49473     Comedy
4    14867      Drama

We could join these tables directly in the database:

In [440]:
pd.read_sql("""
SELECT
    *
FROM
    IMDB
    JOIN genre ON IMDB.Movie_id = genre.Movie_id
""", conn).head(10)

Movie_id                    Title  Rating  TotalVotes MetaCritic  \
0    36809  12 Years a Slave (2013)     8.1      496092         96   
1    36809  12 Years a Slave (2013)     8.1      496092         96   
2    36809  12 Years a Slave (2013)     8.1      496092         96   
3    30114         127 Hours (2010)     7.6      297075         82   
4    30114         127 Hours (2010)     7.6      297075         82   
5    30114         127 Hours (2010)     7.6      297075         82   
6    37367             50/50 (2011)     7.7      283935         72   
7    37367             50/50 (2011)     7.7      283935         72   
8    37367             50/50 (2011)     7.7      283935         72   
9    49473        About Time (2013)     7.8      225412              

       Budget  Runtime CVotes10 CVotes09 CVotes08  ... Votes3044F Votes45A  \
0  20000000.0  134 min    75556   126223   161460  ...        8.0      7.8   
1  20000000.0  134 min    75556   126223   161460  ...        8.0      7.8   
2  20000000.0  134 min    75556   126223   161460  ...        8.0      7.8   
3  18000000.0   94 min    28939    44110    98845  ...        7.5      7.3   
4  18000000.0   94 min    28939    44110    98845  ...        7.5      7.3   
5  18000000.0   94 min    28939    44110    98845  ...        7.5      7.3   
6   8000000.0  100 min    28304    47501    99524  ...        7.6      7.4   
7   8000000.0  100 min    28304    47501    99524  ...        7.6      7.4   
8   8000000.0  100 min    28304    47501    99524  ...        7.6      7.4   
9  12000000.0  123 min    38556    43170    70850  ...        7.7      7.6   

  Votes45AM Votes45AF VotesIMDB Votes1000 VotesUS VotesnUS Movie_id      genre  
0       7.8       8.1       8.0       7.7     8.3      8.0    36809  Biography  
1       7.8       8.1       8.0       7.7     8.3      8.0    36809      Drama  
2       7.8       8.1       8.0       7.7     8.3      8.0    36809    History  
3       7.3       7.5       7.6       7.0     7.7      7.6    30114  Adventure  
4       7.3       7.5       7.6       7.0     7.7      7.6    30114  Biography  
5       7.3       7.5       7.6       7.0     7.7      7.6    30114      Drama  
6       7.4       7.5       7.4       7.0     7.9      7.6    37367     Comedy  
7       7.4       7.5       7.4       7.0     7.9      7.6    37367      Drama  
8       7.4       7.5       7.4       7.0     7.9      7.6    37367    Romance  
9       7.5       7.8       7.7       6.9     7.8      7.7    49473     Comedy  

[10 rows x 54 columns]

Or we could join them in `pandas`.

We need to choose:

- the datasets to join (or "merge" as it's called in `pandas`)
- the column(s) to join on (can be different for each table)
- the type of join (inner, left, right, etc.)

In [441]:
films_merged = films.merge(genres, on="Movie_id", how="inner")

films_merged.head()

Movie_id                    Title  Rating  TotalVotes MetaCritic  \
0    36809  12 Years a Slave (2013)     8.1      496092         96   
1    36809  12 Years a Slave (2013)     8.1      496092         96   
2    36809  12 Years a Slave (2013)     8.1      496092         96   
3    30114         127 Hours (2010)     7.6      297075         82   
4    30114         127 Hours (2010)     7.6      297075         82   

       Budget  Runtime CVotes10 CVotes09 CVotes08  ... Votes3044M Votes3044F  \
0  20000000.0  134 min    75556   126223   161460  ...        7.9        8.0   
1  20000000.0  134 min    75556   126223   161460  ...        7.9        8.0   
2  20000000.0  134 min    75556   126223   161460  ...        7.9        8.0   
3  18000000.0   94 min    28939    44110    98845  ...        7.5        7.5   
4  18000000.0   94 min    28939    44110    98845  ...        7.5        7.5   

  Votes45A Votes45AM Votes45AF VotesIMDB Votes1000 VotesUS VotesnUS      genre  
0      7.8       7.8       8.1       8.0       7.7     8.3      8.0  Biography  
1      7.8       7.8       8.1       8.0       7.7     8.3      8.0      Drama  
2      7.8       7.8       8.1       8.0       7.7     8.3      8.0    History  
3      7.3       7.3       7.5       7.6       7.0     7.7      7.6  Adventure  
4      7.3       7.3       7.5       7.6       7.0     7.7      7.6  Biography  

[5 rows x 53 columns]

Oh oh, looks like our data is duplicated!

In [442]:
print(len(films), len(genres))

117 351


That's exactly 3 genre records per film.

Our options include:

- deduplicating at source (in the SQL query or even the database)
- leaving the data as-is (but we'd have to remember there are 3 rows per film)
- deduplicating *after* the join

Let's see this third option in action:

In [443]:
films_deduped = films_merged.drop_duplicates(subset=["Movie_id"], keep="first")

films_deduped.head()

Movie_id                    Title  Rating  TotalVotes MetaCritic  \
0     36809  12 Years a Slave (2013)     8.1      496092         96   
3     30114         127 Hours (2010)     7.6      297075         82   
6     37367             50/50 (2011)     7.7      283935         72   
9     49473        About Time (2013)     7.8      225412              
12    14867             Amour (2012)     7.9       76121         94   

        Budget  Runtime CVotes10 CVotes09 CVotes08  ... Votes3044M Votes3044F  \
0   20000000.0  134 min    75556   126223   161460  ...        7.9        8.0   
3   18000000.0   94 min    28939    44110    98845  ...        7.5        7.5   
6    8000000.0  100 min    28304    47501    99524  ...        7.6        7.6   
9   12000000.0  123 min    38556    43170    70850  ...        7.6        7.7   
12   8900000.0  127 min    11093    15944    22942  ...        7.7        7.9   

   Votes45A Votes45AM Votes45AF VotesIMDB Votes1000 VotesUS VotesnUS  \
0       7.8       7.8       8.1       8.0       7.7     8.3      8.0   
3       7.3       7.3       7.5       7.6       7.0     7.7      7.6   
6       7.4       7.4       7.5       7.4       7.0     7.9      7.6   
9       7.6       7.5       7.8       7.7       6.9     7.8      7.7   
12      7.9       7.8       8.1       6.6       7.2     7.9      7.8   

        genre  
0   Biography  
3   Adventure  
6      Comedy  
9      Comedy  
12      Drama  

[5 rows x 53 columns]

<h1 style="color: #fcd805">Exercise: combining data</h1>

1. Select all the rows from the `earning` table in the movies database into a `pandas` `DataFrame`.

In [444]:
earning = pd.read_sql("""
SELECT
    *
FROM
    earning
""", conn)

earning.head()


Movie_id  Domestic    Worldwide
0    36809  56671993  187733202.0
1    30114  18335230   60738797.0
2    37367  35014192   39187783.0
3    49473  15322921   87100449.0
4    14867   6739492   19839492.0

2. Now join the earnings data onto the merged film+genre data.

You should now have a `DataFrame` with one row per film and with genre and earnings data added on at the end.

Verify that this is the case before moving on.

In [462]:
ds = films_deduped.merge(earning, on="Movie_id", how="inner")
ds.head()

Movie_id                    Title  Rating  TotalVotes MetaCritic  \
0    36809  12 Years a Slave (2013)     8.1      496092         96   
1    30114         127 Hours (2010)     7.6      297075         82   
2    37367             50/50 (2011)     7.7      283935         72   
3    49473        About Time (2013)     7.8      225412              
4    14867             Amour (2012)     7.9       76121         94   

       Budget  Runtime CVotes10 CVotes09 CVotes08  ... Votes45A Votes45AM  \
0  20000000.0  134 min    75556   126223   161460  ...      7.8       7.8   
1  18000000.0   94 min    28939    44110    98845  ...      7.3       7.3   
2   8000000.0  100 min    28304    47501    99524  ...      7.4       7.4   
3  12000000.0  123 min    38556    43170    70850  ...      7.6       7.5   
4   8900000.0  127 min    11093    15944    22942  ...      7.9       7.8   

  Votes45AF VotesIMDB Votes1000 VotesUS VotesnUS      genre  Domestic  \
0       8.1       8.0       7.7     8.3      8.0  Biography  56671993   
1       7.5       7.6       7.0     7.7      7.6  Adventure  18335230   
2       7.5       7.4       7.0     7.9      7.6     Comedy  35014192   
3       7.8       7.7       6.9     7.8      7.7     Comedy  15322921   
4       8.1       6.6       7.2     7.9      7.8      Drama   6739492   

     Worldwide  
0  187733202.0  
1   60738797.0  
2   39187783.0  
3   87100449.0  
4   19839492.0  

[5 rows x 55 columns]

3. Which film earned the least **domestically**?

In [463]:
ds.sort_values("Domestic").head(1)

Movie_id               Title  Rating  TotalVotes MetaCritic     Budget  \
109    20709  Tyrannosaur (2011)     7.6       26016         65  1000000.0   

    Runtime CVotes10 CVotes09 CVotes08  ... Votes45A Votes45AM Votes45AF  \
109             2060     4083     9078  ...      7.5       7.4       5.8   

    VotesIMDB Votes1000 VotesUS VotesnUS  genre Domestic Worldwide  
109       6.5       7.4     7.6           Drama    22321   22321.0  

[1 rows x 55 columns]

4. Which film earned the most **worldwide**?

In [464]:
ds.sort_values("Worldwide", ascending = False).head(1)

Movie_id                                Title  Rating  TotalVotes  \
79    38626  Star Wars: The Force Awakens (2015)     8.1      676732   

   MetaCritic       Budget  Runtime CVotes10 CVotes09 CVotes08  ... Votes45A  \
79         81  245000000.0  136 min   155391   161810   166378  ...      7.9   

   Votes45AM Votes45AF VotesIMDB Votes1000 VotesUS VotesnUS   genre  \
79       7.8       8.2       8.3       7.7     8.2      7.9  Action   

     Domestic     Worldwide  
79  936662225  2.068224e+09  

[1 rows x 55 columns]

5. How many films have a MetaCritic score of less than 75?

_Note: to answer this question you'll have to fix the data type of the column first and you may need to deal with some non-numeric values!_

In [465]:
ds.head()
ds["MetaCritic"].value_counts()
ds['MetaCritic']=ds['MetaCritic'].replace('', 0)
ds["MetaCritic"].astype(int)
len(ds[ds["MetaCritic"] < 75])
#ds.dtypes

C:\Users\ashut\AppData\Local\Temp\ipykernel_19848\4210365135.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  ds['MetaCritic']=ds['MetaCritic'].replace('', 0)


52

6. Which genre has the highest total domestic earnings?

In [525]:
ds.groupby("genre")["Domestic"].sum().idxmin()

'Mystery'

7. Convert the `Runtime` column to numeric.

_Hint: You'll have to perform some string manipulation on it before you can do this._

In [479]:
ds.head()
ds["Runtime"].value_counts()
ds['Runtime']=ds['Runtime'].replace('', "0 min")
ds["Runtime"].value_counts()
##ds["Runtime"].str.split(" ")[0] --- ask this 
copyds=ds
ds["Runtime_mins"] = ds["Runtime"].str[:-4]
ds["Runtime_mins"] = ds["Runtime_mins"].astype(int)
ds.dtypes


Movie_id         object
Title            object
Rating          float64
TotalVotes        int64
MetaCritic        int64
Budget           object
Runtime          object
CVotes10         object
CVotes09         object
CVotes08         object
CVotes07         object
CVotes06         object
CVotes05         object
CVotes04         object
CVotes03         object
CVotes02         object
CVotes01         object
CVotesMale       object
CVotesFemale     object
CVotesU18        object
CVotesU18M       object
CVotesU18F       object
CVotes1829       object
CVotes1829M      object
CVotes1829F      object
CVotes3044       object
CVotes3044M      object
CVotes3044F      object
CVotes45A        object
CVotes45AM       object
CVotes45AF       object
CVotes1000       object
CVotesUS         object
CVotesnUS        object
VotesM           object
VotesF           object
VotesU18         object
VotesU18M        object
VotesU18F        object
Votes1829        object
Votes1829M       object
Votes1829F      

8. Now find the genre with the highest **median** runtime.

In [482]:
ds.groupby(ds["genre"])["Runtime_mins"].median().idxmax()

'Mystery'

<h1 style="color: #fcd805">Exercise: pub names</h1>

Let's do some open-ended data analysis with `pandas`!

We're going to find out what the most common pub name is in the UK.

1. Read in the file `open_pubs.csv` from the `data` folder into a `pandas` `DataFrame` (data originally from https://www.getthedata.com/open-pubs).

In [493]:
pubs = pd.read_csv("data/open_pubs.csv", header=None)
header = pd.read_csv("data/myheaders.txt",sep="|")
header_list = header["Field"][1:]
#print(header_list)
#header.head()
pubs.columns = header_list
pubs.head()

Field  fsa_id                name  \
0          22          Anchor Inn   
1          36  Ark Bar Restaurant   
2          74           Black Boy   
3          75         Black Horse   
4          76          Black Lion   

Field                                            address  postcode  easting  \
0            Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW   604749   
1      Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH   610194   
2      The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA   587334   
3            Lower Street, Stratford St Mary, COLCHESTER   CO7 6JS   622675   
4                          Lion Road, Glemsford, SUDBURY  CO10 7RF   622675   

Field  northing   latitude longitude local_authority  
0        234404  51.970379  0.979340         Babergh  
1        233329  51.958698  1.057832         Babergh  
2        241316  52.038595  0.729915         Babergh  
3      -5527598         \N        \N         Babergh  
4      -5527598         \N        \N         Babergh

2. Looks like there are no column headers!

Here is the data dictionary for the dataset:

|Field|Data type|Comments|
|---|---|---|
|fsa_id|int|Food Standard Agency's ID for this pub.|
|name|string|Name of the pub.|
|address|string|Address fields separated by commas.|
|postcode|string|Postcode of the pub.|
|easting|int| |
|northing|int| |
|latitude|decimal| |
|longitude|decimal| |
|local_authority|string|Local authority this pub falls under.|

Read the documentation for the `read_csv` method and figure out how to add column names to the data when you read it in.

3. Check for any missing data. Drop any row with no name, since we need values from that column.

In [497]:
pubs.isnull().sum()

Field
fsa_id             0
name               0
address            0
postcode           0
easting            0
northing           0
latitude           0
longitude          0
local_authority    0
dtype: int64

4. Convert the `name` column (or whatever you called it) to the correct `string` type.

In [528]:
pubs.dtypes
pubs["name"] = pubs["name"].astype('string')
pubs.dtypes

Field
fsa_id                      int64
name               string[python]
address                    object
postcode                   object
easting                     int64
northing                    int64
latitude                   object
longitude                  object
local_authority            object
dtype: object

5. Now convert the values in the `name` column to lowercase so that names like "The King's Arms" and "The king's arms" are treated as the same name.

In [505]:
pubs["name"]=pubs["name"].str.lower().str.strip().str.replace("the ","")


Field  fsa_id                name  \
0          22          anchor inn   
1          36  ark bar restaurant   
2          74           black boy   
3          75         black horse   
4          76          black lion   
...       ...                 ...   
51306  596947            swan inn   
51307  596975    bersham club ltd   
51308  596995         cross foxes   
51309  596998          drunk monk   
51310  597004           four dogs   

Field                                            address  postcode  easting  \
0            Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW   604749   
1      Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH   610194   
2      The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA   587334   
3            Lower Street, Stratford St Mary, COLCHESTER   CO7 6JS   622675   
4                          Lion Road, Glemsford, SUDBURY  CO10 7RF   622675   
...                                                  ...       ...      ...   
51306  Swan Inn Mount Street, Rhostyllen, Wrexham, Wr...  LL14 4AU   331271   
51307  Bersham Sports And Social Club Bersham Road, B...  LL14 4HS   331318   
51308     Last Orders, 15 Abbot Street, Wrexham, Wrexham  LL11 1TA   333434   
51309                 3 Overton Arcade, Wrexham, Wrexham  LL13 8HR   333572   
51310           The Four Dogs Box Lane, Wrexham, Wrexham  LL12 8EF   333764   

Field  northing   latitude  longitude local_authority  
0        234404  51.970379   0.979340         Babergh  
1        233329  51.958698   1.057832         Babergh  
2        241316  52.038595   0.729915         Babergh  
3      -5527598         \N         \N         Babergh  
4      -5527598         \N         \N         Babergh  
...         ...        ...        ...             ...  
51306    348880  53.032774  -3.026330         Wrexham  
51307    349279  53.036358  -3.025714         Wrexham  
51308    350218  53.045074  -2.994358         Wrexham  
51309    350192  53.044849  -2.992294         Wrexham  
51310    351861  53.059882  -2.989775         Wrexham  

[51311 rows x 9 columns]

6. Use the `.str.strip()` method to remove any trailing whitespace from the `name` column.

7. Now use `.str.replace` to remove the word "the" from the pub names, so that a pub called "The King's Head" will be treated as having the same name as one that's simply called "King's Head".

*Tip: take care not to replace words that **contain** the word `the` like "theatre"*

8. Use your `name` column to find the most common pub name in the UK.

In [513]:
pubs["name"].value_counts().idxmax()

'red lion'

BONUS: which local authority has the most of these pubs (i.e. the most pubs that have the most common name you found in question 8)?

In [529]:
pubs[pubs["name"]=="red lion"]["local_authority"].value_counts().idxmax()

local_authority
West Northamptonshire    10
Buckinghamshire          10
East Lindsey              8
Cherwell                  7
Wiltshire                 7
                         ..
Gwynedd                   1
Neath Port Talbot         1
Merthyr Tydfil            1
Newport                   1
Swansea                   1
Name: count, Length: 165, dtype: int64

BONUS: how many unique pub names are there in the data? That is, pub names that appear exactly once.

In [534]:
count_pubs = pubs.groupby("name")["name"].count().sort_values()
count_pubs[count_pubs==1]

name
zombie shack                            1
 half moon                              1
 horseshoes - three horseshoes          1
 lion                                   1
 regent bar                             1
                                       ..
west bridgford conservative club        1
west bridgford lawn tennis club         1
west bromwich albion supporters club    1
west bromwich dartmouth cricke          1
west bulls                              1
Name: name, Length: 30728, dtype: int64

# `pandas` help

- pandastutor visualises what different operations do: https://pandastutor.com/
- `pandas` cheat sheet: https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf
- Python for Data Analysis book (free online) by the creator of `pandas`: https://wesmckinney.com/book/